data-pipeline/src/pipeline/main.py

```diff
@@ -437,61 +470,119 @@ def pre_process_bfr(run_type, year):
+def _get_ancillary_data(
+    worker_client: Client,
+    run_id: str,
+    year: int,
+) -> tuple[pd.DataFrame]:
+    """
+    Retrieve and process supporting data files.
+
+    Note: `run_type` is _always_ "default".
+
+    :param worker_client: Dask client
+    :param run_id: unique identifier (used for write target)
+    :param year: financial year in question
+    :return: tuple of supporting datasets
+    """
+    run_type = "default"
 
     cdc, census, sen, ks2, ks4, aar, schools, cfo, central_services = (
         worker_client.gather(
             [
-                worker_client.submit(pre_process_cdc, run_type, year),
-                worker_client.submit(pre_process_census, run_type, year),
-                worker_client.submit(pre_process_sen, run_type, year),
-                worker_client.submit(pre_process_ks2, run_type, year),
-                worker_client.submit(pre_process_ks4, run_type, year),
-                worker_client.submit(pre_process_academy_ar, run_type, year),
-                worker_client.submit(pre_process_schools, run_type, year),
-                worker_client.submit(pre_process_cfo, run_type, year),
-                worker_client.submit(pre_process_central_services, run_type, year),
+                worker_client.submit(pre_process_cdc, run_type, year, run_id),
+                worker_client.submit(pre_process_census, run_type, year, run_id),
+                worker_client.submit(pre_process_sen, run_type, year, run_id),
+                worker_client.submit(pre_process_ks2, run_type, year, run_id),
+                worker_client.submit(pre_process_ks4, run_type, year, run_id),
+                worker_client.submit(pre_process_academy_ar, run_type, year, run_id),
+                worker_client.submit(pre_process_schools, run_type, year, run_id),
+                worker_client.submit(pre_process_cfo, run_type, year, run_id),
+                worker_client.submit(
+                    pre_process_central_services, run_type, year, run_id
+                ),
             ]
         )
     )
 
-    data_ref = worker_client.scatter(
+    return worker_client.scatter(
         (schools, census, sen, cdc, aar, ks2, ks4, cfo, central_services)
     )
 
+
+def pre_process_data(
+    worker_client: Client,
+    run_id: str,
+    aar_year: int,
+    cfr_year: int,
+    bfr_year: int,
+):
+    """
+    Process data necessary for Academies, Maintained Schools BFR and Trusts.
+
+    Note: `run_type` is _always_ "default".
+
+    :param worker_client: Dask client
+    :param run_id: unique identifier (used for write target)
+    :param aar_year: Academy financial year/source
+    :param cfr_year: Maintained School financial year/source
+    :param bfr_year: BFR year/source
+    :return: duration of processing
+    """
+    run_type = "default"
+
+    start_time = time.time()
+    logger.info(f"Pre-processing data {run_type} - {run_id}")
+
+    academy_data_ref = maintained_data_ref = _get_ancillary_data(
```

Comment:
@jrabbott, as discussed, this is the bit in question:

- if `aar_year` and `cfr_year` are the same, both effectively have the same set of data
- if `cfr_year` is different (functionally, this will always mean _later_), it'll repeat this in its entirety looking exclusively at the `cfr_year` directory